installer/block_device.go

```diff
@@ -195,3 +252,155 @@ func (bd *BlockDevice) SectorSize() (int, error) {
+	// TODO -- Add nil guards
+
+	log.Infof("blockFrameWriter: len(b []byte): %d. b: %s", len(b), string(b[:2]))
+
+	// Fill the frame buffer first
+	n, err = bw.buf.Write(b)
+	if err != nil {
+		log.Errorf("bd.buf.Write: %s", err.Error())
+		return n, err
+	}
+
+	if bw.buf.Len() < bw.frameSize {
+		log.Infof("Fewer than chunksize bytes written")
+		return n, nil // Chunk buffer not full
+	}
+
+	log.Info("BlockFrameWriter: Flushing buffer")
+	totWritten := 0
+	nFrames := bw.buf.Len() / bw.frameSize
+	for i := 0; i < nFrames; i++ {
+		n, err = bw.w.Write(bw.buf.Next(bw.frameSize))
+		totWritten += n
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Report the cached bytes as written
+	totWritten += bw.buf.Len() % bw.frameSize
+
+	// Write a full frame, but report only the last byte chunk as written
+	return len(b), nil
+}
+
+// Close flushes the remaining cached bytes, if any.
+func (bw *BlockFrameWriter) Close() error {
+	_, err := bw.w.Write(bw.buf.Bytes())
+	if cerr := bw.w.Close(); cerr != nil {
+		return cerr
+	}
+	return err
+}
+
+// OptimizedBlockDeviceWriter wraps an underlying blockDevice write, however,
+// with the optimization that it compares the bytes passed in to the Write
+// method, with the next len([]byte) bytes (considered a frame) on the block
+// device, and if they match, the write is discarded. The lingo is that only
+// dirty frames are written. Clean ones are discarded.
+type OptimizedBlockDeviceWriter struct {
+	blockDevice BlockDevicer
+	totalFrames int
+	dirtyFrames int
+}
+
+// Write only write 'dirty' frames.
+// Note: a frame-size is always the size 'len(b)'
+func (bd *OptimizedBlockDeviceWriter) Write(b []byte) (n int, err error) {
+
+	log.Infof("Optimized block device writer: len(b): %d, b: %s", len(b), string(b[:2]))
+	log.Infof("Framesize: %d", len(b))
+
+	frameSize := int64(len(b))
+	payloadBuf := make([]byte, frameSize)
+
+	//
+	// Read len(b) bytes from the block-device
+	//
+	n, err = io.ReadFull(bd.blockDevice, payloadBuf)
+	if err != nil {
+		log.Errorf("Failed to read a full frame of size: %d from the block-device", err)
+		return 0, err
+	}
+
+	//
+	// Write the frame if it is dirty.
+	//
+	if !bytes.Equal(payloadBuf, b) {
+		log.Info("Dirty frame detected. Rewriting...")
+		// In order to write, we need to seek back to
+		// the start of the chunk.
+		if _, err = bd.blockDevice.Seek(-int64(frameSize), io.SeekCurrent); err != nil {
+			log.Errorf("Failed to seek back to the start of the frame")
+			return 0, err
+		}
+		bd.totalFrames += 1
+		bd.dirtyFrames += 1
+		return bd.blockDevice.Write(b)
+	}
+
+	log.Infof("Skipping writing of clean frame. Reporting: %d bytes as written", n)
+	// No need to write a clean frame
+	bd.totalFrames += 1
+	return n, err
+}
+
+func (obw *OptimizedBlockDeviceWriter) Close() error {
+	s := "The optimized block-device writer wrote a total of %d frames, " +
+		"where %d frames did need to be rewritten"
+	log.Infof(s, obw.totalFrames, obw.dirtyFrames)
```

Comment:
Nice statistics! :+1:

Reply:
Yeh, nice idea. I'm struggling a bit with the wording though.